packages/contractkit/src/wallets/signers/azure-key-vault-client.ts

```diff
@@ -0,0 +1,191 @@
+    if (!(await this.hasKey(keyName))) {
+      throw new Error(`Unable to locate key: ${keyName}`)
+    }
+    return (await this.getKey(keyName)).id!
+  }
+
+  public async signMessage(message: Buffer, keyName: string): Promise<Signature> {
+    if (!(await this.hasKey(keyName))) {
+      throw new Error(`Unable to locate key: ${keyName}`)
+    }
+    const keyId = await this.getKeyId(keyName)
+    const cryptographyClient = new CryptographyClient(keyId, this.credential)
+    // @ts-ignore-next-line (ECDSA256 is not included in the client enum but is valid)
+    const signResult = await cryptographyClient.sign(this.SIGNING_ALGORITHM, message)
+    // The output of this will be a 64 byte array.
+    // The first 32 are the value for R and the rest is S
+    if (
+      typeof signResult === 'undefined' ||
+      typeof signResult.result === 'undefined' ||
+      signResult.result.length !== 64
+    ) {
+      throw new Error(`Invalid signature returned from Azure: ${signResult}`)
+    }
+    const rawSignature = signResult.result
+
+    // Canonicalize signature
+    const secp256k1Curve = new EC('secp256k1')
+    const R = AzureKeyVaultClient.bufferToBigNumber(Buffer.from(rawSignature.slice(0, 32)))
+    let S = AzureKeyVaultClient.bufferToBigNumber(Buffer.from(rawSignature.slice(32, 64)))
+
+    // The Azure Signature MAY not be canonical, which is illegal in Ethereum
+    // thus it must be transposed to the lower intersection.
+    // https://github.com/bitcoin/bips/blob/master/bip-0062.mediawiki#Low_S_values_in_signatures
+    const N = AzureKeyVaultClient.bufferToBigNumber(secp256k1Curve.curve.n)
+    if (!AzureKeyVaultClient.isCanonical(S, N)) {
+      S = secp256k1Curve.curve.n.sub(S)
+    }
+
+    const rBuff = AzureKeyVaultClient.bigNumberToBuffer(R)
+    const sBuff = AzureKeyVaultClient.bigNumberToBuffer(S)
+    const canonicalizedSignature = Buffer.concat([rBuff, sBuff])
+    const publicKey = await this.getPublicKey(keyName)
+
+    // Azure doesn't provide the recovery key in the signature
+    const recoveryParam = AzureKeyVaultClient.recoverKeyIndex(
+      canonicalizedSignature,
+      publicKey,
+      message
+    )
+    return new Signature(recoveryParam, rBuff, sBuff)
+  }
+
+  public async hasKey(keyName: string): Promise<boolean> {
+    try {
+      await this.keyClient.getKey(keyName)
+    } catch (e) {
+      if (e.message.includes('this is not a valid private key')) {
+        return false
+      }
+      throw e
+    }
+    return true
+  }
+
+  /**
+   * Returns true if the signature is in the "bottom" of the curve
+   */
+  private static isCanonical(S: BigNumber, curveN: BigNumber): boolean {
+    return S.comparedTo(curveN.shiftedBy(1)) <= 0
+  }
+
+  /**
+   * Attempts each recovery key to find a match
+   */
+  private static recoverKeyIndex(
+    signature: Uint8Array,
+    publicKey: BigNumber,
+    hash: Uint8Array
+  ): number {
+    for (let i = 0; i < 4; i++) {
+      const compressed = false
```

Comment:
nit: no need for this (unless if you want it for readability?)

Reply:
This was intentional to be self-documenting. It doesn't look like TS supports named parameters.